### PR TITLE
feat(cmd): Adding new logging and Tree command and adding test to check for duplicate commands

### PR DESCRIPTION
**Pull request by arehmandev**
Features:
- New command `corectl tenants tree` - displays tenancy tree
- Logs are now captured in ~/.config/corectl/log.json

Fix:
- Removed duplicate corectl commands and added test to avoid in future

### CODE REVIEW

**Comment by Malet** [comment]
Good find! I think the runtime checks may be unnecessary here though, https://github.com/coreeng/corectl/pull/52/files#diff-4a2124744a6aaf77fd02653dc11bd0b4b683aa79e22fa37d6f286bf4aefc4785L138-L140 would probably suffice for this PR

File: pkg/cmd/tenant/tenant.go [hunk 1]
```diff
@@ -26,6 +27,7 @@ func NewTenantCmd(cfg *config.Config) *cobra.Command {
 	tenantCmd.AddCommand(describe.NewTenantDescribeCmd(cfg))
 	tenantCmd.AddCommand(addrepo.NewTenantAddRepoCmd(cfg))
 	tenantCmd.AddCommand(create.NewTenantCreateCmd(cfg))
+	tenantCmd.AddCommand(tree.NewTenantsTreeCmd(cfg))
```
Comment (Malet) [inline pkg/cmd/tenant/tenant.go:30]: Need to add in the history command!

File: pkg/cmd/root/root_test.go [hunk 2]
```diff
@@ -26,3 +32,81 @@ func TestRootSetsValidLogLevel(t *testing.T) {
...
 }
+
+func TestCheckDuplicateCmds(t *testing.T) {
+	log.DefaultLogger.SetLevel(log.Level(8))
+	config := config.NewConfig()
+	fillCfgWithMockValues(config)
+	rootCmd := NewRootCmd(config)
+
+	// Check there's no duplicate commands
+	cmdErr := checkDuplicateCmds(rootCmd)
```
Comment (Malet) [inline pkg/cmd/root/root_test.go:43]: I like that this is now in the tests :+1:

File: pkg/history/history.go [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+		select {
+		case <-done:
+		case <-sigChan:
+		}
+
+		// Log help command execution
+		logEntry := CommandLog{
+			Date:        time.Now(),
+			Input:       strings.Join(os.Args, " "),
+			Output:      stdout.String(),
```
Comment (Malet) [inline pkg/history/history.go:202]: Do we know why some commands end up with nil output?:

```json
{
  "date": "2024-11-04T16:02:37.962884629Z",
  "input": "./corectl app create js-app --log-level debug --dry-run",
  "exitCode": 0,
  "command_type": "APPS_CREATE"
},
{
  "date": "2024-11-04T16:40:31.492094867Z",
  "input": "./corectl update",
  "exitCode": 130,
  "command_type": "UPDATE"
}
```

File: pkg/history/history.go [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+				return err
+			}
+		} else if originalPreRun != nil {
+			originalPreRun(cmd, args)
+		}
+		return nil
+	}
+
+	// Handle RunE commands (commands that can return errors)
+	if cmd.RunE != nil {
```
Comment (Malet) [inline pkg/history/history.go:226]: Could we try to dry up the function bodies for `RunE` vs `Run` they seem to do ~80% of the same thing

File: pkg/history/history.go (hunk 3 above)
Comment (arehmandev) [reply pkg/history/history.go:202]: Interactive commands I believe, will take a look if theres anything i can do

File: pkg/cmd/root/root_test.go [hunk 5]
```diff
@@ -26,3 +32,81 @@ func TestRootSetsValidLogLevel(t *testing.T) {
 	ConfigureGlobalLogger("panic")
 	assert.Equal(t, log.PanicLevel, log.DefaultLogger.Level)
 }
+
+func TestCheckDuplicateCmds(t *testing.T) {
+	log.DefaultLogger.SetLevel(log.Level(8))
```
Comment (dark0dave) [inline pkg/cmd/root/root_test.go:37]: Would a setup tear down be appropriate here? You are modifying the default logger, may want to set it back after you have touched it.

Comment (dark0dave) [reply pkg/cmd/root/root_test.go:37]: https://pkg.go.dev/testing#T.Cleanup

File: pkg/cmd/root/root_test.go [hunk 6]
```diff
@@ -26,3 +32,85 @@ func TestRootSetsValidLogLevel(t *testing.T) {
...
+	// Remove duplicate nested commands again
+	rootCmd.Commands()[0].RemoveCommand((duplicateCommands[0]))
+	rootCmd.Commands()[0].RemoveCommand((duplicateCommands[0]))
+	cmdErr = checkDuplicateCmds(rootCmd)
+	assert.Nil(t, cmdErr)
+}
+
+// A simple recursive check to just make sure commands and subcommands are unique
+func checkDuplicateCmds(rootCmd *cobra.Command) (err error) {
+	var duplicateCmds []string
```
Comment (dark0dave) [inline pkg/cmd/root/root_test.go:85]: This array is not needed.

File: pkg/cmd/root/root_test.go [hunk 7]
```diff
@@ -26,3 +32,85 @@ func TestRootSetsValidLogLevel(t *testing.T) {
 	ConfigureGlobalLogger("panic")
 	assert.Equal(t, log.PanicLevel, log.DefaultLogger.Level)
 }
+
+func TestCheckDuplicateCmds(t *testing.T) {
```
Comment (dark0dave) [inline pkg/cmd/root/root_test.go:36]: This test is long feels like there is a better way you can handle this. I'd break it up into multiple tests, something like:
- Test plain duplicate commands
- Test nested

File: pkg/cmd/tenant/tree/tenant_tree.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+				newPrefix += "    "
+			} else {
+				newPrefix += vertical
+			}
+			t.writeChildren(sb, node.Children, newPrefix)
+		}
+	}
+}
+
+func printTreeView(cfg *config.Config) error {
```
Comment (dark0dave) [inline pkg/cmd/tenant/tree/tenant_tree.go:136]: should this be part of the Tree struct? 

ie 

func (t *tree) printTreeView ?

File: pkg/cmd/tenant/tree/tenant_tree_test.go [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	"testing"
+
+	"github.com/coreeng/developer-platform/pkg/tenant"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCreateTreeView(t *testing.T) {
+
+	var tenants []tenant.Tenant
+	tenants = append(tenants, tenant.Tenant{Name: "root"})
```
Comment (dark0dave) [inline pkg/cmd/tenant/tree/tenant_tree_test.go:13]: consider:

```go
tenants := []{...}
```

File: pkg/history/history.go [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+func getLogPath() (string, error) {
+	// Get user's home directory
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return "", fmt.Errorf("could not get home directory: %w", err)
+	}
+
+	configDir := filepath.Join(home, logFilePath)
+
+	// Create directory structure if it doesn't exist
```
Comment (dark0dave) [inline pkg/history/history.go:68]: If you feel the need to comment this, perhaps create a function instead.

File: pkg/cmd/root/root_test.go (hunk 6 above)
Comment (arehmandev) [reply pkg/cmd/root/root_test.go:85]: General guidelines is to report all failures rather than failfast

Comment (dark0dave) [reply pkg/cmd/root/root_test.go:85]: Sure, then return the hashmap.

Do either one of the following:

- Return hashmap, remove array
- Remove array, print when you have the same command twice (you know from the hashmap)

File: pkg/logger/logger.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+func Init() error {
+	if Log != nil {
+		return nil
+	}
+
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		fmt.Printf("Failed to get home directory: %v\n", err)
+		os.Exit(1)
```
Comment (dark0dave) [inline pkg/logger/logger.go:46]: What? Why? What do you need from the user's home dir?|

Comment (dark0dave) [reply pkg/logger/logger.go:46]: This will not work in containers.

Comment (dark0dave) [reply pkg/logger/logger.go:46]: pick a better location for logs.

Comment (dark0dave) [reply pkg/logger/logger.go:46]: The same dir as thie binary or

```sh
/var/log
```

File: pkg/logger/logger.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	LogPath = strings.Replace(LogFile, "$HOME", homeDir, 1)
+
+	logDir := filepath.Dir(LogPath)
+	if err := os.MkdirAll(logDir, 0755); err != nil {
```
Comment (dark0dave) [inline pkg/logger/logger.go:51]: please no,don't make executable

**Review by dark0dave — CHANGES_REQUESTED** [review]
Split out the tree and history stuff.
Don't log to home dir.
logging changes are fine otherwise.

File: pkg/logger/logger.go (hunk 11 above)
Comment (arehmandev) [reply pkg/logger/logger.go:46]: It was on the ticket, we use ~/.config/corectl folder throughout

File: pkg/logger/logger.go (hunk 12 above)
Comment (arehmandev) [reply pkg/logger/logger.go:51]: 755 is pretty standard for a folder?
